File: src/main/java/LC0002_AddTwoNumbers/AddTwoNumbers.java
package LC0002_AddTwoNumbers;

class Solution {

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode l3 = new ListNode();
        ListNode p = l1;
        ListNode q = l2;
        ListNode r = l3;

        int carry = 0;
        while (p != null || q != null || carry > 0) {
            int sum = carry;
            if (p != null) {
                sum += p.val;
            }
            if (q != null) {
                sum += q.val;
            }
            carry = sum / 10;

            r.next = new ListNode(sum % 10);
            r = r.next;
            if (p != null) {
                p = p.next;
            }
            if (q != null) {
                q = q.next;
            }
        }

        return l3.next;
    }
}

File: LeetCode/cpp/LC0002_Add_Two_Numbers/AddTwoNumbers.cpp
#include <iostream>
#include "ListNode.h"

using namespace std;

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *res = new ListNode;

        int carry = 0;
        ListNode *p1 = l1, *p2 = l2, *r = nullptr;
        while (p1 || p2) {
            if (p1) {
                carry += p1->val;
                p1 = p1->next;
            }
            if (p2) {
                carry += p2->val;
                p2 = p2->next;
            }
            ListNode *s = new ListNode;
            s->next = nullptr;
            s->val = carry % 10;
            carry /= 10;

            if (!r) {
                res = s;
            } else {
                r->next = s;
            }
            r = s;
        }

        if (carry) {
            ListNode *s = new ListNode;
            s->next = nullptr;
            s->val = carry;
            r->next = s;
        }

        return res;
    }
};